### PR TITLE
Improve library loading time by ~3.5x

### PR DESCRIPTION
**Pull request by mborsetti**
## Proposed change

This PR (an advanced beta) introduces a working architecture that improves the load speed of the library by ~3.5x, addressing all issues in #998.

Besides the significant speed improvement, it fixes several other things:
* Eliminates object naming confusion -- makes all objects created by this library of the same class, simplifying debugging for developers and making the IDE documentation useful;
* Checks validity of country code -- prompts users to contribute to the library if we do not have holidays for a valid country.
* Reduce bloat (future) -- we can remove the hundreds of duplicate class names and keep only a single one per country/market;
* Improves documentation.

The architecture schematics is as follows:
* The `country_holidays` and `market_holidays` loading functions have been upgraded to know what countries and markets we support, and to instantiate a new, unified, class Holidays.
* This new class Holidays will monkeypatch itself with the attributes and methods contained in the country-specific class, after loading the submodule for that country or market. **It will not load any other submodule (country or market) than the one needed**.
* The library's `_init__.py` has been upgraded to detect the way the library is being imported: 
   * If it is imported using e.g. `import holidays`, it does not change the behavior and slowly imports all objects ensuring backward compatibility. 
   * If the import is e.g. `from holidays import country_holidays` then it will not import those objects to allow for fast lazy loading through the function/Holiday class.

This PR is beta (and **limited to Python >= 3.11**) because I would like a review and approval before I go forward with cleaning it up:
* Fixing `pre-commit-config.yaml` to work on Windows, where there is no `make`;
* Thorough testing on previous Python versions and backporting it to as many previous Python versions as practical;
* Write specific tests for the new functionality (e.g. whether entering a valid ISO 3166-1 alpha-2 code for a country that we do not support raises a warning (actually, exception!) to invite the user to contribute to the library and so on) and ensure 100% coverage;
* Upgrade more existing tests so they instantiate the holiday object with the default instantiation functions instead of loading a submodule directly;
* Do more docstring/documentation cleanup/improvements, including full removal of ``expand`` as we 100% **do not use** (separate PR?);
* Hunt for other improvements by studying the `importtime` trace below;
* Fix `isort` and `black` undoing each other's changes;
* Rebasing to latest beta code;
* etc.

Here are the `importtime` startup traces from my machine, where loading improves from ~245216 to ~89633 (~3.5x):

**Before**
```
> python -X importtime -c 'import holidays'
import time: self [us] | cumulative | imported package
import time:       138 |        138 | winreg
import time:       321 |        321 |   _io
import time:        49 |         49 |   marshal
import time:       236 |        236 |   nt
import time:       914 |       1518 | _frozen_importlib_external
import time:      1624 |       1624 |   time
import time:       336 |       1960 | zipimport
import time:        60 |         60 |     _codecs
import time:       396 |        455 |   codecs
import time:      1182 |       1182 |   encodings.aliases
import time:      2053 |       3689 | encodings
import time:       656 |        656 | encodings.utf_8
import time:        51 |         51 | _signal
import time:        37 |         37 |     _abc
import time:       207 |        243 |   abc
import time:       242 |        484 | io
import time:        49 |         49 |       _stat
import time:       159 |        207 |     stat
import time:       645 |        645 |     _collections_abc
import time:        38 |         38 |       genericpath
import time:        74 |         74 |       _winapi
import time:       746 |        857 |     ntpath
import time:       479 |       2186 |   os
import time:        75 |         75 |   _sitebuiltins
import time:       562 |        562 |   encodings.cp1252
import time:      1623 |       1623 |   _distutils_hack
import time:       652 |        652 |   types
import time:       640 |        640 |       warnings
import time:       680 |       1320 |     importlib
import time:       617 |        617 |     importlib._abc
import time:       766 |        766 |         itertools
import time:      1415 |       1415 |         keyword
import time:       404 |        404 |           _operator
import time:      1666 |       2070 |         operator
import time:      1121 |       1121 |         reprlib
import time:       133 |        133 |         _collections
import time:      2437 |       7940 |       collections
import time:       172 |        172 |         _functools
import time:      1233 |       1405 |       functools
import time:       981 |      10326 |     contextlib
import time:       383 |      12645 |   importlib.util
import time:        97 |         97 |   importlib.machinery
import time:       223 |        223 |   sitecustomize
import time:       363 |        363 |   usercustomize
import time:     11449 |      29871 | site
import time:       846 |        846 |       _ast
import time:      1811 |       1811 |       enum
import time:      1792 |       4448 |     ast
import time:        47 |         47 |         _opcode
import time:       987 |       1034 |       opcode
import time:      1479 |       2513 |     dis
import time:       725 |        725 |     collections.abc
import time:       225 |        225 |             _sre
import time:       895 |        895 |               re._constants
import time:      1128 |       2023 |             re._parser
import time:       581 |        581 |             re._casefix
import time:      1697 |       4524 |           re._compiler
import time:       740 |        740 |           copyreg
import time:      1628 |       6891 |         re
import time:       842 |        842 |         token
import time:      2128 |       9860 |       tokenize
import time:      1709 |      11569 |     linecache
import time:      3434 |      22687 |   inspect
import time:        79 |         79 |     _typing
import time:      3231 |       3309 |   typing
import time:       793 |        793 |         _weakrefset
import time:      1048 |       1840 |       weakref
import time:       256 |        256 |           org
import time:       145 |        401 |         org.python
import time:       152 |        553 |       org.python.core
import time:      1042 |       3434 |     copy
import time:       112 |        112 |         math
import time:       129 |        129 |         _datetime
import time:      2745 |       2985 |       datetime
import time:       139 |        139 |         _locale
import time:     17470 |      17608 |       locale
import time:      1752 |      22344 |     calendar
import time:      1693 |       1693 |     gettext
import time:       347 |        347 |           posix
import time:       338 |        684 |         posixpath
import time:      1309 |       1993 |       fnmatch
import time:       118 |        118 |       errno
import time:       871 |        871 |         urllib
import time:      2357 |       3227 |       urllib.parse
import time:      2184 |       7520 |     pathlib
import time:      1008 |       1008 |         dateutil._version
import time:      1707 |       2714 |       dateutil
import time:       817 |        817 |         __future__
import time:        61 |         61 |           _string
import time:      1571 |       1632 |         string
import time:        73 |         73 |             _struct
import time:       768 |        840 |           struct
import time:      1815 |       2655 |         six
import time:       767 |        767 |             numbers
import time:      1743 |       2509 |           _decimal
import time:       858 |       3367 |         decimal
import time:       397 |        397 |           dateutil._common
import time:       814 |       1210 |         dateutil.relativedelta
import time:       108 |        108 |               _bisect
import time:      1728 |       1835 |             bisect
import time:        70 |         70 |             six.moves
import time:      1046 |       1046 |             dateutil.tz._common
import time:      1514 |       1514 |             dateutil.tz._factories
import time:      4365 |       4365 |                 _ctypes
import time:       871 |        871 |                 ctypes._endian
import time:      2020 |       7255 |               ctypes
import time:       992 |        992 |               ctypes.wintypes
import time:      1163 |       9410 |             dateutil.tz.win
import time:      3989 |      17861 |           dateutil.tz.tz
import time:       762 |      18623 |         dateutil.tz
import time:      2179 |      30481 |       dateutil.parser._parser
import time:      1088 |       1088 |       dateutil.parser.isoparser
import time:      1643 |      35926 |     dateutil.parser
import time:       867 |        867 |     holidays.constants
import time:      2351 |      74133 |   holidays.holiday_base
import time:       458 |        458 |     dateutil.easter
import time:      1060 |       1518 |   holidays.financial.european_central_bank
import time:       415 |        415 |           hijri_converter.helpers
import time:       686 |        686 |           hijri_converter.locales
import time:       573 |        573 |           hijri_converter.ummalqura
import time:      1161 |       2834 |         hijri_converter.convert
import time:       818 |       3651 |       hijri_converter
import time:       782 |       4433 |     holidays.calendars
import time:       708 |       5140 |   holidays.financial.ny_stock_exchange
import time:       797 |        797 |   holidays.countries.albania
import time:       586 |        586 |     holidays.countries.united_states
import time:       649 |       1235 |   holidays.countries.american_samoa
import time:       517 |        517 |   holidays.countries.andorra
import time:       567 |        567 |   holidays.countries.angola
import time:      1186 |       1186 |   holidays.countries.argentina
import time:       744 |        744 |   holidays.countries.armenia
import time:       565 |        565 |   holidays.countries.aruba
import time:       696 |        696 |   holidays.countries.australia
import time:       668 |        668 |   holidays.countries.austria
import time:       657 |        657 |   holidays.countries.azerbaijan
import time:       620 |        620 |   holidays.countries.bahrain
import time:       528 |        528 |   holidays.countries.bangladesh
import time:       483 |        483 |   holidays.countries.belarus
import time:       477 |        477 |   holidays.countries.belgium
import time:       487 |        487 |   holidays.countries.bolivia
import time:       518 |        518 |   holidays.countries.bosnia_and_herzegovina
import time:       560 |        560 |   holidays.countries.botswana
import time:       508 |        508 |   holidays.countries.brazil
import time:       475 |        475 |   holidays.countries.bulgaria
import time:       492 |        492 |   holidays.countries.burundi
import time:       498 |        498 |   holidays.countries.canada
import time:       629 |        629 |       pymeeus
import time:       437 |        437 |       pymeeus.base
import time:       697 |        697 |       pymeeus.Angle
import time:      1179 |       2941 |     pymeeus.Epoch
import time:       537 |        537 |         pymeeus.Interpolation
import time:      1126 |       1663 |       pymeeus.Coordinates
import time:      2111 |       2111 |       pymeeus.Earth
import time:       940 |       4713 |     pymeeus.Sun
import time:      1150 |       1150 |         sysconfig
import time:      1153 |       2302 |       zoneinfo._tzpath
import time:       547 |        547 |       zoneinfo._common
import time:       761 |        761 |       _zoneinfo
import time:      1095 |       4705 |     zoneinfo
import time:       997 |      13354 |   holidays.countries.chile
import time:       528 |        528 |   holidays.countries.china
import time:       523 |        523 |   holidays.countries.colombia
import time:       508 |        508 |   holidays.countries.croatia
import time:       514 |        514 |   holidays.countries.cuba
import time:       508 |        508 |   holidays.countries.curacao
import time:       501 |        501 |   holidays.countries.cyprus
import time:       502 |        502 |   holidays.countries.czechia
import time:       495 |        495 |   holidays.countries.denmark
import time:       815 |        815 |   holidays.countries.djibouti
import time:      2190 |       2190 |   holidays.countries.dominican_republic
import time:      1620 |       1620 |   holidays.countries.egypt
import time:      1292 |       1292 |   holidays.countries.estonia
import time:       985 |        985 |   holidays.countries.eswatini
import time:      1151 |       1151 |   holidays.countries.ethiopia
import time:       687 |        687 |   holidays.countries.finland
import time:       869 |        869 |   holidays.countries.france
import time:       566 |        566 |   holidays.countries.georgia
import time:       545 |        545 |   holidays.countries.germany
import time:       488 |        488 |   holidays.countries.greece
import time:       502 |        502 |   holidays.countries.guam
import time:       644 |        644 |   holidays.countries.honduras
import time:       543 |        543 |   holidays.countries.hongkong
import time:       710 |        710 |   holidays.countries.hungary
import time:       868 |        868 |   holidays.countries.iceland
import time:       928 |        928 |   holidays.countries.india
import time:       717 |        717 |   holidays.countries.indonesia
import time:       919 |        919 |   holidays.countries.ireland
import time:       554 |        554 |     holidays.countries.united_kingdom
import time:      1137 |       1690 |   holidays.countries.isle_of_man
import time:       376 |        376 |           convertdate.utils
import time:       557 |        932 |         convertdate.gregorian
import time:       517 |        517 |         convertdate.julian
import time:       949 |       2397 |       convertdate.armenian
import time:       523 |        523 |       convertdate.bahai
import time:       402 |        402 |       convertdate.coptic
import time:       382 |        382 |         convertdate.julianday
import time:       572 |        954 |       convertdate.daycount
import time:       382 |        382 |       convertdate.dublin
import time:       401 |        401 |           convertdate.data
import time:      1066 |       1467 |         convertdate.data.french_republican_days
import time:       640 |       2107 |       convertdate.french_republican
import time:      1078 |       1078 |       convertdate.hebrew
import time:       508 |        508 |         convertdate.islamic
import time:      1245 |       1752 |       convertdate.holidays
import time:       509 |        509 |       convertdate.indian_civil
import time:       476 |        476 |       convertdate.iso
import time:       802 |        802 |       convertdate.mayan
import time:       572 |        572 |       convertdate.persian
import time:       592 |        592 |         convertdate.data.positivist
import time:       569 |       1160 |       convertdate.positivist
import time:       421 |        421 |       convertdate.ordinal
import time:      2652 |      16183 |     convertdate
import time:       617 |      16799 |   holidays.countries.israel
import time:       501 |        501 |   holidays.countries.italy
import time:       428 |        428 |   holidays.countries.jamaica
import time:       528 |        528 |   holidays.countries.japan
import time:       423 |        423 |   holidays.countries.kazakhstan
import time:       454 |        454 |   holidays.countries.kenya
import time:       452 |        452 |   holidays.countries.kyrgyzstan
import time:       419 |        419 |   holidays.countries.latvia
import time:       420 |        420 |   holidays.countries.lesotho
import time:       441 |        441 |   holidays.countries.liechtenstein
import time:       419 |        419 |   holidays.countries.lithuania
import time:       398 |        398 |   holidays.countries.luxembourg
import time:       424 |        424 |   holidays.countries.madagascar
import time:       419 |        419 |   holidays.countries.malawi
import time:       515 |        515 |   holidays.countries.malaysia
import time:       510 |        510 |   holidays.countries.malta
import time:       798 |        798 |   holidays.countries.marshall_islands
import time:       587 |        587 |   holidays.countries.mexico
import time:       669 |        669 |   holidays.countries.moldova
import time:       638 |        638 |   holidays.countries.monaco
import time:       492 |        492 |   holidays.countries.montenegro
import time:       533 |        533 |   holidays.countries.morocco
import time:       749 |        749 |   holidays.countries.mozambique
import time:       541 |        541 |   holidays.countries.namibia
import time:       506 |        506 |   holidays.countries.netherlands
import time:       581 |        581 |   holidays.countries.new_zealand
import time:       575 |        575 |   holidays.countries.nicaragua
import time:       589 |        589 |   holidays.countries.nigeria
import time:       581 |        581 |   holidays.countries.north_macedonia
import time:       526 |        526 |   holidays.countries.northern_mariana_islands
import time:        54 |         54 |         _heapq
import time:       636 |        690 |       heapq
import time:      1138 |       1828 |     dateutil.rrule
import time:       627 |       2454 |   holidays.countries.norway
import time:       554 |        554 |   holidays.countries.pakistan
import time:       437 |        437 |   holidays.countries.panama
import time:       519 |        519 |   holidays.countries.paraguay
import time:       468 |        468 |   holidays.countries.peru
import time:       455 |        455 |   holidays.countries.philippines
import time:       459 |        459 |   holidays.countries.poland
import time:       477 |        477 |   holidays.countries.portugal
import time:       442 |        442 |   holidays.countries.puerto_rico
import time:       441 |        441 |   holidays.countries.romania
import time:       440 |        440 |   holidays.countries.russia
import time:       439 |        439 |   holidays.countries.san_marino
import time:       442 |        442 |   holidays.countries.saudi_arabia
import time:       444 |        444 |   holidays.countries.serbia
import time:      1172 |       1172 |   holidays.countries.singapore
import time:      1568 |       1568 |   holidays.countries.slovakia
import time:       752 |        752 |   holidays.countries.slovenia
import time:       582 |        582 |   holidays.countries.south_africa
import time:       915 |        915 |       korean_lunar_calendar.korean_lunar_calendar
import time:       811 |       1725 |     korean_lunar_calendar
import time:       769 |       2494 |   holidays.countries.south_korea
import time:       586 |        586 |   holidays.countries.spain
import time:       470 |        470 |   holidays.countries.sweden
import time:       498 |        498 |   holidays.countries.switzerland
import time:       458 |        458 |   holidays.countries.taiwan
import time:       946 |        946 |   holidays.countries.thailand
import time:       895 |        895 |   holidays.countries.tunisia
import time:       656 |        656 |   holidays.countries.turkey
import time:       535 |        535 |   holidays.countries.ukraine
import time:       493 |        493 |   holidays.countries.united_arab_emirates
import time:       451 |        451 |   holidays.countries.united_states_minor_outlying_islands
import time:       433 |        433 |   holidays.countries.united_states_virgin_islands
import time:       649 |        649 |   holidays.countries.uruguay
import time:       975 |        975 |   holidays.countries.uzbekistan
import time:       855 |        855 |   holidays.countries.vatican_city
import time:       606 |        606 |   holidays.countries.venezuela
import time:      1038 |       1038 |   holidays.countries.vietnam
import time:       788 |        788 |   holidays.countries.zambia
import time:       556 |        556 |   holidays.countries.zimbabwe
import time:     30211 |     245216 | holidays
```

**After**
```
> python -X importtime -c 'from holidays import country_holidays'
import time: self [us] | cumulative | imported package
import time:       112 |        112 | winreg
import time:       204 |        204 |   _io
import time:        53 |         53 |   marshal
import time:       162 |        162 |   nt
import time:       756 |       1173 | _frozen_importlib_external
import time:      1223 |       1223 |   time
import time:       272 |       1494 | zipimport
import time:        61 |         61 |     _codecs
import time:       429 |        490 |   codecs
import time:      1488 |       1488 |   encodings.aliases
import time:      1944 |       3921 | encodings
import time:       973 |        973 | encodings.utf_8
import time:        68 |         68 | _signal
import time:        43 |         43 |     _abc
import time:       303 |        346 |   abc
import time:       345 |        690 | io
import time:        61 |         61 |       _stat
import time:       215 |        275 |     stat
import time:       794 |        794 |     _collections_abc
import time:        36 |         36 |       genericpath
import time:        74 |         74 |       _winapi
import time:       800 |        910 |     ntpath
import time:       495 |       2473 |   os
import time:        66 |         66 |   _sitebuiltins
import time:       726 |        726 |   encodings.cp1252
import time:      1702 |       1702 |   _distutils_hack
import time:       766 |        766 |   types
import time:       743 |        743 |       warnings
import time:       817 |       1559 |     importlib
import time:       718 |        718 |     importlib._abc
import time:       188 |        188 |         itertools
import time:      1095 |       1095 |         keyword
import time:       173 |        173 |           _operator
import time:      1758 |       1930 |         operator
import time:       904 |        904 |         reprlib
import time:        95 |         95 |         _collections
import time:      2634 |       6844 |       collections
import time:       130 |        130 |         _functools
import time:      1189 |       1319 |       functools
import time:      1381 |       9543 |     contextlib
import time:       580 |      12399 |   importlib.util
import time:        89 |         89 |   importlib.machinery
import time:       225 |        225 |   sitecustomize
import time:       263 |        263 |   usercustomize
import time:     11585 |      30290 | site
import time:       853 |        853 |       _ast
import time:      1729 |       1729 |       enum
import time:      1773 |       4354 |     ast
import time:        53 |         53 |         _opcode
import time:      1022 |       1075 |       opcode
import time:      1431 |       2505 |     dis
import time:       682 |        682 |     collections.abc
import time:       303 |        303 |             _sre
import time:       802 |        802 |               re._constants
import time:      1597 |       2399 |             re._parser
import time:       545 |        545 |             re._casefix
import time:      1798 |       5044 |           re._compiler
import time:       705 |        705 |           copyreg
import time:      1978 |       7726 |         re
import time:       730 |        730 |         token
import time:      2945 |      11400 |       tokenize
import time:       796 |      12195 |     linecache
import time:      3501 |      23237 |   inspect
import time:        44 |         44 |     _typing
import time:      2767 |       2810 |   typing
import time:       730 |        730 |         _weakrefset
import time:      1027 |       1756 |       weakref
import time:       247 |        247 |           org
import time:       142 |        388 |         org.python
import time:       196 |        584 |       org.python.core
import time:      1051 |       3390 |     copy
import time:        68 |         68 |         math
import time:       109 |        109 |         _datetime
import time:      1903 |       2080 |       datetime
import time:        58 |         58 |         _locale
import time:      1427 |       1485 |       locale
import time:      1396 |       4960 |     calendar
import time:      1146 |       1146 |     gettext
import time:       303 |        303 |           posix
import time:       222 |        525 |         posixpath
import time:       917 |       1441 |       fnmatch
import time:       123 |        123 |       errno
import time:       938 |        938 |         urllib
import time:      2149 |       3087 |       urllib.parse
import time:      1456 |       6106 |     pathlib
import time:       674 |        674 |         dateutil._version
import time:       963 |       1637 |       dateutil
import time:       635 |        635 |         __future__
import time:        63 |         63 |           _string
import time:      1635 |       1697 |         string
import time:       127 |        127 |             _struct
import time:      1353 |       1480 |           struct
import time:      2887 |       4366 |         six
import time:      1420 |       1420 |             numbers
import time:      2428 |       3848 |           _decimal
import time:      1957 |       5804 |         decimal
import time:       603 |        603 |           dateutil._common
import time:       904 |       1507 |         dateutil.relativedelta
import time:        91 |         91 |               _bisect
import time:      1230 |       1320 |             bisect
import time:       126 |        126 |             six.moves
import time:      1231 |       1231 |             dateutil.tz._common
import time:       851 |        851 |             dateutil.tz._factories
import time:      3449 |       3449 |                 _ctypes
import time:       937 |        937 |                 ctypes._endian
import time:      1469 |       5854 |               ctypes
import time:      1036 |       1036 |               ctypes.wintypes
import time:       908 |       7797 |             dateutil.tz.win
import time:      2070 |      13393 |           dateutil.tz.tz
import time:       729 |      14122 |         dateutil.tz
import time:      2393 |      30522 |       dateutil.parser._parser
import time:       735 |        735 |       dateutil.parser.isoparser
import time:       997 |      33889 |     dateutil.parser
import time:       437 |        437 |     holidays.constants
import time:      1989 |      51915 |   holidays.holiday_base
import time:     11672 |      89633 | holidays
```

## Type of change

- [ ] New country/market holidays support (thank you!)
- [ ] Supported country/market holidays update (calendar discrepancy fix, localization)
- [x] Existing code/test/process quality improvement (best practice, cleanup, refactoring, optimization)
- [ ] Dependency upgrade (version update)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (a code change causing existing functionality to break)
- [x] New feature (new python-holidays functionality in general)

## Checklist

- [x] I've followed the [contributing guidelines][contributing-guidelines]
- [x] This PR is filed against `beta` branch of the repository
- [ ] This PR doesn't contain any merge conflicts and has clean commit history
- [x] The code style looks good (`make pre-commit`)
- [ ] I've added tests to verify that the new code works and all tests pass locally (`make test`, `make tox`)
- [x] The related [documentation][docs] has been added/updated (check off the box for free if no updates is required)

[contributing-guidelines]: https://github.com/dr-prodigy/python-holidays/blob/beta/CONTRIBUTING.rst
[docs]: https://github.com/dr-prodigy/python-holidays/tree/beta/docs/source

### CODE REVIEW

File: holidays/utils.py [hunk 1]
```diff
@@ -175,85 +437,398 @@ def country_holidays(
...
+    # organisations per https://en.wikipedia.org/wiki/ISO_3166-2:RS#Note
+    iso_3166_1.add("XK")
+
+    country = country.upper()
+    if country not in iso_3166_1:
+        raise ValueError(
+            f"'{country}' is not an ISO 3166-1 alpha-2 code; see "
+            f"https://en.wikipedia.org/wiki/ISO_3166-1#Current_codes."
+        )
+    if country not in country_to_module:
```
Comment (KJhellico) [inline holidays/utils.py:723]: 👍

**Review by KJhellico — COMMENTED** [review]
@mborsetti, it looks very impressive!

**Comment by arkid15r** [comment]
Looks massive :)
I'm going to look at it over the weekend. Need to play with it locally for sure.

Thanks for submitting this Mike!

File: holidays/countries/albania.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ def _add_with_observed(
                     + td(days=+2 if self._is_saturday(hol_date) else days)
                 ] = f"{hol_name} (Observed)"
 
-        super()._populate(year)
+        HolidayBase._populate(self, year)
```
[the remark this answers is not in the document]
Comment (mborsetti) [reply holidays/countries/albania.py:44]: Actually it's because when it's monkey patched into the Holidays class, the method loses the reference as to what is considered `super()`. 

A stylistic cost of backwards compatibility!

File: holidays/countries/singapore.py [hunk 3]
```diff
@@ -81,30 +81,32 @@ def __init__(
         See parameters and usage in :py:class:`HolidayBase`.
         """
 
-        self.cnls = _ChineseLuniSolar()
```
[the remark this answers is not in the document]
Comment (mborsetti) [reply holidays/countries/singapore.py:84]: Agree, and that's why I wrote Singapore that way, but for some reason this assignment didn't monkey patch into Holidays. Since (1) it appeared only once and (2) I wrote the code for Singapore, I took the liberty of changing it for now. 

Will put figuring out a way to monkey patch in the todo.

**Comment by mborsetti** [comment]
> I'm going to look at it over the weekend. Need to play with it locally for sure.

Look forward to the comments! 

@arkid15r @KJhellico I will not do any work until I have a clear go ahead.

File: tox.ini [hunk 4]
```diff
@@ -19,7 +19,7 @@ deps =
     -r{toxinidir}/requirements/dev.txt
     -r{toxinidir}/requirements/docs.txt
 commands =
-    make doc
+    sphinx-build -E -T -W -b html -D language=en -j auto -q docs/source docs/build  # make not on Windows
```
Comment (arkid15r) [inline tox.ini:22]: https://stackoverflow.com/questions/2532234/how-to-run-a-makefile-in-windows may help with this.

**Review by arkid15r — COMMENTED** [review]
Thanks for the huge effort, Mike. I really enjoyed reviewing this PR. And it IS massive :).

Its massiveness is not the biggest thing that concerns me as we've been able to handle quite deep changes in recent releases. I'm concerned with the compatibility (our major priority) issues this PR would introduce. You can find my comments below and the general idea is that we don't want to lose ability to import country/financial entities right from holidays module.

Another point is complexity -- I'm not saying the code is bad structured or hard to read (your comments helped a lot). I wasn't familiar with some `inspect` stuff and in general I feel it can be done in a simpler manner. The PR is really well-documented and also introduces some great ideas which I'd love PH to have implemented.

The bottom line: I guess the ideal option here would be to work together on a separate PR that combines simpler lazy import approach and the best parts of the current PR. Let me know if that's  something you would consider.

It'd be also great to hear other contributors/participants/watchers opinions as I may be missing or misunderstanding some points.

Thank you!

**Comment by mborsetti** [comment]
> You can find my comments below and the general idea is that we don't want to lose ability to import country/financial entities right from holidays module.

Hmmm... not sure if/how this can be achievable. The extreme slowness in loading is largely due to everything being accessible (and therefore loaded) from the top level at all times; I look forward to seeing your solution.

> Another point is complexity -- I'm not saying the code is bad structured or hard to read (your comments helped a lot). I wasn't familiar with some `inspect` stuff and in general I feel it can be done in a simpler manner. The PR is really well-documented and also introduces some great ideas which I'd love PH to have implemented.

Hmmm, is the issue complexity or your lack of knowledge of a specific Python module? In any way I would love to see a simpler solution!!   IMHO the complexity unfortunately comes from having to pay for the sins of the original architecture: a breaking implementation would be overall far simpler and less bloated than the current one, not just 3.5x faster.

> The bottom line: I guess the ideal option here would be to work together on a separate PR that combines simpler lazy import approach and the best parts of the current PR. Let me know if that's something you would consider.

Absolutely!!  Yet not at all sure as to why it has to be a _separate_ PR?

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -32,10 +32,13 @@ repos:
     rev: v5.10.1
     hooks:
       - id: isort
-        # holidays/countries/__init__.py is excluded due to a bug whereby
-        # black and isort don't agree and enter into an infinite fix loop.
+        # holidays/countries/__init__.py is excluded due to imports being in
+        # relevance order vs. alphabetical one.
+        # tests/countries/test_united_states_virgin_islands.py is excluded due
+        # to a bug whereby black and isort don't agree and enter into an
```
[the remark this answers is not in the document]
Comment (mborsetti) [reply .pre-commit-config.yaml:38]: Could be the Windows vs Linux implementation of the packages, but I did experience it!

File: docs/source/examples.rst [hunk 6]
```diff
@@ -8,10 +8,10 @@ Simplest example possible
 .. code-block:: python
 
    >>> from datetime import date
-   >>> import holidays
-   >>> date(2014, 1, 1) in holidays.US()
+   >>> from holidays import country_holidays
```
[the remark this answers is not in the document]
Comment (mborsetti) [reply docs/source/examples.rst:11]: No, it doesn't at all. That's the backward compatibility part.

File: holidays/utils.py [hunk 7]
```diff
@@ -13,114 +13,370 @@
...
 from typing import Dict, Iterable, List, Optional, Union
 
-from holidays import countries, financial
 from holidays.holiday_base import HolidayBase
 
+country_to_module = {
+    # List of all countries we support and the submodule name
+    # The code itself must be the name of the class in that submodule
+    # All country codes are ISO 3166-1 alpha-2 codes
+    "AL": "albania",
```
[the remark this answers is not in the document]
Comment (mborsetti) [reply holidays/utils.py:32]: I assume you meant alpha-2, and yes.

File: tox.ini (hunk 4 above)
Comment (mborsetti) [reply tox.ini:22]: That's a heavy hack, installing all sort of non-native bloat. Sorry but no thanks!

File: holidays/utils.py [hunk 8]
```diff
@@ -13,114 +13,370 @@
...
+    perform argument validation and provide documentation, as follows:
+
+    >>> from holidays import country_holidays
+    >>> us_holidays = country_holidays('US')
+
+    >>> from holidays import financial_holidays
+    >>> nyse_holidays = financial_holidays('XNYS')
+    """
+
+    def __init__(
```
[the remark this answers is not in the document]
Comment (mborsetti) [reply holidays/utils.py:196]: `expand` is supported in the public API to maintain 100% backwards compatibility. It's streamlined away internally since it's implemented in exactly zero lines of code for zero countries or financial markets.

File: .pre-commit-config.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (KJhellico) [reply .pre-commit-config.yaml:38]: I always had this issue too on Windows (with Python 3.11). Just check on Ubuntu (with Python 3.8) - the same, `isort` splits a long multiline import string into two shorter ones. With isort 5.12.0 (latest from https://github.com/pycqa/isort) - the same.

[the remark this answers is not in the document]
Comment (arkid15r) [reply .pre-commit-config.yaml:38]: Yeah, our current black/isort config is a bit "unorthodox". It seems it needs some attention.

**Comment by arkid15r** [comment]
> You can still do `from holidays import US`, it's just 3.5x slower. This PR gives 100% backwards compatibility.

It didn't work in an expected way for me (it still isn't) so that lead me to this question (see `import holidays` example code above).

> Hmmm, is the issue complexity or your lack of knowledge of a specific Python module? In any way I would love to see a simpler solution!! IMHO the complexity unfortunately comes from having to pay for the sins of the original architecture: a breaking implementation would be overall far simpler and less bloated than the current one, not just 3.5x faster, and the `inspect` bit is to pay for the sins of enabling imports such as `from holidays import US` and so on.

My inexperience with these particular methods doesn't make the code less complex. It makes it harder to figure that out for me. In a long run it'll need to be maintained and might require some extra effort.
The architectural constraints are definitely the culprit and this makes the task challenging and interesting. I'm going to look into alternative implementation and let you know / request a review next week.

> Absolutely!! Yet not at all sure as to why it has to be a _separate_ PR?

I see it as a work that should be done in a "from scratch" fashion. I would definitely build it upon current PR if I wanted to improve something based on this PR code. You can find examples how I used this approach [here](https://github.com/dr-prodigy/python-holidays/pull/889) or [here](https://github.com/dr-prodigy/python-holidays/pull/888).

Let me know if I missed something based on your replies.
I'm looking forward to getting this task done together within the next two weeks.

Thanks Mike!

**Comment by mborsetti** [comment]
> It didn't work in an expected way for me (it still isn't) so that lead me to this question (see `import holidays` example code above)

Gotcha.

> I see it as a work that should be done in a "from scratch" fashion. I would definitely build it upon current PR if I wanted to improve something based on this PR code. You can find examples how I used this approach [here](https://github.com/dr-prodigy/python-holidays/pull/889) or [here](https://github.com/dr-prodigy/python-holidays/pull/888).

Sounds good. 

Look forward to seeing what you come up with!
